src/lowering/desugar.ml

```diff
@@ -90,8 +90,8 @@ and exp' at note = function
       (breakE "!" (nullE()))
       (* case ? v : *)
       (varP v) (varE v) ty).it
-  | S.ObjBlockE (s, _t, dfs) ->
-    obj_block at s None dfs note.Note.typ
+  | S.ObjBlockE (s, (self_id_opt,_), dfs) ->
```

Comment:
This was the bug, which was failing to introduce a the self declaration for some cases.